### PR TITLE
Failsafe fixes

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
Add changes proposed in https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/5083

#### Change overview
- arm failsafe on PASE session establishment completion
- do not allow commissioning window to be opened if the failsafe is held
- expire failsafe on commissioning window close
- do not allow CASE sessions to arm the failsafe for the first time if the commissioning window is open

#### Testing
Please see new cirque tests. Tests for arming failsafe on PASE session establishment completion are more complicated because we have no direct access to the failsafe value. This can be tested by attempting to send a command that requires the failsafe to be armed. Network commands, unfortunately, do not yet check this. I have added such a check in https://github.com/cecille/connectedhomeip/tree/failsafe_fixes_with_tests along with the tests, but failsafe checking on network commands needs to be handled in another PR as it may affect platform devs.

### CODE REVIEW

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp [hunk 1]
```diff
@@ -165,10 +167,33 @@ bool emberAfGeneralCommissioningClusterArmFailSafeCallback(app::CommandHandler *
...
     {
-        CheckSuccess(failSafeContext.ArmFailSafe(accessingFabricIndex, System::Clock::Seconds16(commandData.expiryLengthSeconds)),
-                     Failure);
-        response.errorCode = CommissioningError::kOk;
-        commandObj->AddResponse(commandPath, response);
+        // We do not allow CASE connections to arm the failsafe for the first time while the commissioning window is open in order
+        // to allow commissioners the opportunity to obtain this failsafe for the purpose of commissioning
+        if (!failSafeContext.IsFailSafeArmed() &&
+            Server::GetInstance().GetCommissioningWindowManager().CommissioningWindowStatus() !=
+                app::Clusters::AdministratorCommissioning::CommissioningWindowStatus::kWindowNotOpen &&
```
Comment (bzbarsky-apple) [inline src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:174]:
```suggestion
                AdministratorCommissioning::CommissioningWindowStatus::kWindowNotOpen &&
```

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp [hunk 2]
```diff
@@ -165,10 +167,33 @@ bool emberAfGeneralCommissioningClusterArmFailSafeCallback(app::CommandHandler *
...
+            commandObj->GetExchangeContext()->GetSessionHandle()->GetSessionType() == Transport::Session::SessionType::kSecure &&
+            commandObj->GetExchangeContext()->GetSessionHandle()->AsSecureSession()->GetSecureSessionType() ==
+                Transport::SecureSession::Type::kCASE)
```
Comment (bzbarsky-apple) [inline src/app/clusters/general-commissioning-server/general-commissioning-server.cpp:177]: Might be simpler as:
```suggestion
            commandObj->GetSubjectDescriptor().authMode == Access::AuthMode::kCase)
```
but either way.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp (hunk 1 above)
Comment (cecille) [reply src/app/clusters/general-commissioning-server/general-commissioning-server.cpp]: Added this manually in the next commit so I could test.

File: src/app/clusters/general-commissioning-server/general-commissioning-server.cpp (hunk 2 above)
Comment (cecille) [reply src/app/clusters/general-commissioning-server/general-commissioning-server.cpp]: Added this manually in the next commit so I could test.

**Review by msandstedt — APPROVED** [review]

**Pull request merged**